edx-solutions/discussion-edx-platform-extensions
# [MCKIN-8418] Optimize UsersSocialMetrics.get API

This optimizes getting user's engagement score by storing his stats in this Django app instead of calculating them every time the change is made.


`paver test_system -s lms -t social_engagement` and `./manage.py lms --settings test test edx_solutions_api_integration` run tests for this one.

**Before testing**
Checkout the `api_integration` and `edx-platform` to `agrendalath/MCKIN-8418-optimize-social-metrics`.
Set `'ENABLE_SOCIAL_ENGAGEMENT': True` in `/lms/envs/common.py`.

Endpoint for getting stats:
```bash
curl -X GET \
  'http://localhost:8000/api/server/users/{USER_ID}/courses/{COURSE_ID}/metrics/social/?include_stats=true' \
  -H 'Cache-Control: no-cache' \
  -H 'X-Edx-Api-Key: edx_api_key'
```

**Manual testing instructions**:
1. Prepare two users.
2. Create some threads, check `num_threads`.
3. Create some comments in thread, check `num_comments`, `num_comments_generated`.
4. Create some replies for comment, check `num_replies`, `num_comments_generated`.
5. Upvote some threads and comments as the second user, check `num_upvotes`, remove vote, check `num_upvotes`.
6. Follow and some threads as the second user, check `num_thread_followers`, unfollow some threads, check `num_thread_followers`.
7. Flag some posts, check `num_flagged`.
8. Delete some replies, check `num_replies` and `num_flagged`.
9. Delete some comments with replies, check `num_comments`, `num_replies`, `num_upvotes` and `num_flagged`.
10. Delete some threads with nested comments and replies, check `num_threads`, `num_comments_generated`, `num_comments`, `num_replies`, `num_upvotes`, `num_flagged` and `num_thread_followers`.
